lib/app_profiler/server.rb

```diff
@@ -0,0 +1,292 @@
+          @transport = ProfileServer::ProfileServerTransportUNIX.new
+        when TRANSPORT_TCP
+          @transport = ProfileServer::ProfileServerTransportTCP.new(AppProfiler::Server.port)
+        else
+          raise "invalid transport #{transport}"
+        end
+
+        @listen_thread = nil
+
+        AppProfiler.logger.info(
+          "[AppProfiler::Server] listening on addr=#{@transport.socket.addr}"
+        )
+        at_exit { stop }
+      end
+
+      def client
+        @transport.client
+      end
+
+      def serve
+        return unless @listen_thread.nil?
+
+        app = ProfileApplication.new
+
+        @listen_thread = Thread.new do
+          loop do
+            Thread.new(@transport.socket.accept) do |session|
+              request = session.gets
+              if request.nil?
+                session.close
+                next
+              end
+              method, path, http_version = request.split(" ")
+              path_info, query_string = path.split("?")
+              env = { # an extremely minimal rack env hash, just enough to get the job done
+                "HTTP_VERSION" => http_version,
+                "REQUEST_METHOD" => method,
+                "PATH_INFO" => path_info,
+                "QUERY_STRING" => query_string,
+                "rack.input" => "",
+              }
+              status, headers, body = app.call(env)
+
+              begin
+                session.print("#{http_version} #{status}\r\n")
+                headers.each do |header, value|
+                  session.print("#{header}: #{value}\r\n")
+                end
+                session.print("\r\n")
+                body.each do |part|
+                  session.print(part)
+                end
+              rescue => e
+                AppProfiler.logger.error(
+                  "[AppProfiler::Server] exception #{e} responding to request #{request}: #{e.message}"
+                )
+              ensure
+                session.close
+              end
+            end
+          end
+        end
+      end
+
+      def stop
+        @listen_thread.kill
+        @transport.stop
+      end
+    end
+
+    class << self
+      def reset!
+        @profile_server = {}
+        DEFAULTS.each do |config, value|
+          class_variable_set(:"@@#{config}", value) # rubocop:disable Style/ClassVars
+        end
+      end
+
+      def start!
+        return if profile_server.key?(Process.pid)
+
+        profile_server[Process.pid] = ProfileServer.new(AppProfiler::Server.transport)
+        profile_server[Process.pid].serve
+        profile_server[Process.pid]
+      end
+
+      def client
+        return unless profile_server.key?(Process.pid)
+
+        profile_server[Process.pid].client
+      end
+
+      def stop!
+        return unless profile_server.key?(Process.pid)
+
+        profile_server[Process.pid].stop
+        profile_server.delete(Process.pid)
+      end
+
+      private
+
+      def profile_server
```

Comment:
Nit: Generally I would expect `profile_servers` to equal `{}` and `profile_server` to be `profile_servers[Process.pid]`. Would the hash here actually be shared among processes though? What is the scenario where we expect there to be more than one server?

Reply:
good point, i'll update this 👍

Reply:
Updated and added a comment to clarify the reason for the hash